File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpDatabase.java
package com.tcssol.expensetracker.Data;


import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.tcssol.expensetracker.Model.PersonExp;
import com.tcssol.expensetracker.Utils.Converters;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
@Database(entities = {PersonExp.class},version = 1,exportSchema = false)
@TypeConverters({Converters.class})
public abstract class PersonExpDatabase extends RoomDatabase {
    public static final int NUMBER_OF_THREADS=4;
    public static final String DATABASE_NAME="person_exp_database2";
    private static volatile PersonExpDatabase INSTANCE;
    public static final ExecutorService databaseWriterExecutor
            = Executors.newFixedThreadPool(NUMBER_OF_THREADS);
    public static final RoomDatabase.Callback sRoomDatabaseCallback=
            new RoomDatabase.Callback(){
                @Override
                public void onCreate(@NonNull SupportSQLiteDatabase db) {
                    super.onCreate(db);
                    databaseWriterExecutor.execute(()->{
                        //invoke Dao, and write
                        PersonExpDao expenseDao=INSTANCE.personExpDao();
                        expenseDao.deleteAll();// clean slate

                        //writing to our table
                    });
                }
            };

    public static PersonExpDatabase getDatabase(final Context context){
        if(INSTANCE==null){
            synchronized (PersonExpDatabase.class){
                if(INSTANCE==null){
                    INSTANCE= Room.databaseBuilder(context.getApplicationContext(),
                                    PersonExpDatabase.class,DATABASE_NAME)
                            .addCallback(sRoomDatabaseCallback)
                            .allowMainThreadQueries()
                            .build();
                }
            }
        }
        return INSTANCE;
    }
    public abstract PersonExpDao personExpDao();

}

File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpDatabase.kt
package com.tcssol.expensetracker.Data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import com.tcssol.expensetracker.Model.PersonExp
import com.tcssol.expensetracker.Utils.Converters
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@Database(entities = [PersonExp::class], version = 2, exportSchema = false)
@TypeConverters(Converters::class)
abstract class PersonExpDatabase : RoomDatabase() {

    abstract fun personExpDao(): PersonExpDao

    companion object {
        const val DATABASE_NAME = "person_exp_database2"

        @Volatile
        private var INSTANCE: PersonExpDatabase? = null

        val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE person_expenses ADD COLUMN note TEXT")
            }
        }

        private val sRoomDatabaseCallback = object : RoomDatabase.Callback() {
            override fun onCreate(db: SupportSQLiteDatabase) {
                super.onCreate(db)
                INSTANCE?.let { database ->
                    CoroutineScope(Dispatchers.IO).launch {
                        database.personExpDao().deleteAll()
                    }
                }
            }
        }

        fun getDatabase(context: Context): PersonExpDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    PersonExpDatabase::class.java,
                    DATABASE_NAME
                )
                    .addCallback(sRoomDatabaseCallback)
                    .addMigrations(MIGRATION_1_2)
                    .build()
                INSTANCE = instance
                instance
            }
        }
    }
}
